### ISSUE TITLE
Experiment: deterministic max-len A/B to isolate target truncation

### ISSUE THREAD

**Issue by TeaShaman-cyber** [TeaShaman-cyber/theseus-needle-lab#8]
## Research question

Does removing training-target truncation improve Needle 2 microlearning when every other controllable factor is held fixed?

This experiment is a deterministic A/B diagnostic. It is not a scale-up, an Orchard/teacher-trajectory test, or a claim of generalization.

## Why this experiment exists

Issue #4 established that the original 12-row, one-epoch LoRA is behaviorally active but fails held-out and training replay. Subsequent probes ruled out decode budget as the main cause and showed export quantization is only secondary.

A 2026-08-31 training-contract audit found that the original `--max-len 128` truncates two rows in the actual training set, both inside the JSON tool-call target. The same run also has only five non-zero-LR parameter updates and an unseeded epoch shuffle.

The first new experiment should isolate truncation before increasing training strength.

## Fixed lineage

Use the same public 12-row dataset and pinned implementation lineage already recorded by Issues #1/#4:

```text
cactus-needle 2.0.8
upstream reference: cactus-compute/needle@ee221ce7c13579d9809209b979a9b7a50936614c
base checkpoint SHA256: 4b0a972d163ffc7678fb3c36bace508114872e9d2ce9e10f225825752d3795bc
```

## Paired arms

Both arms must run in separate fresh processes with the same explicit NumPy seed before invoking the pinned fine-tune implementation.

Common configuration:

```text
same 12 rows
same deterministic validation split
same seeded epoch shuffle
same epochs=1
same batch_size=2
same lr=1e-4
same LoRA rank=4
same LoRA alpha=32
same no-generate path
same W4 export
same inference max_new_tokens=256
same train-replay and held-out evaluators
```

Only intentional difference:

```text
A: max_len cap = 128
B: max_len cap = 1024 (upstream default; actual fit bucket is derived by Needle)
```

## Required pre-training evidence

Before either fine-tune executes, write a machine-readable target-coverage report using the installed Needle tokenizer/render contract. It must record per row:

- prompt tokens;
- target tokens;
- effective sequence length;
- target tokens retained;
- whether the full target and EOS are retained;
- validation/training membership.

The workflow must fail rather than silently proceed if arm B still truncates a training target.

## Measurements

For each arm:

- exact adapter and W4 artifact hashes;
- fine-tune logs, wall time, peak RSS;
- reported step count/loss;
- exact 12-row train replay;
- existing 24-row held-out evaluation;
- confusion counts and changed predictions;
- target-coverage report;
- explicit seed/config in a machine-readable receipt.

## Interpretation

```text
B materially improves train replay over A
-> target truncation is causally implicated in this bounded setup

A and B remain near the same poor train replay
-> truncation is not the dominant limitation; next probe may vary training strength
```

Held-out results are secondary until the model can first express the training task.

## Boundaries

- Do not change epochs, LR, rank, alpha, dataset, or teacher supervision in this experiment.
- Do not merge Orchard Issue #6 into this experiment.
- Green CI means the experiment executed as specified; scientific disposition comes from the measured A/B result.

**Comment by TeaShaman-cyber** [comment]
## Deterministic max-len A/B — VERIFIED NEGATIVE 2026-08-31

The paired experiment specified by this Issue completed and was independently recomputed from the downloaded GitHub artifact.

### Lineage

```text
branch  experiment/needle-maxlen-ab
initial experiment commit 84f62209cff7c1d903daa4aad601ea935fae5cdf
runtime-evidence correction c6664ad4f22c25c358316aae97e7850f69279471
PR #9
```

The first run `33428115215` stopped before training because the pre-training gate correctly rejected a stale expected truncation set `[7,9,12]`. Authoritative runtime tokenizer evidence showed the exact set is `[7,9]`; row 12 fits exactly at 128 tokens with EOS retained. Published audit records were corrected rather than preserving the stale reconstruction as FACT.

### Successful exact-SHA run

```text
run        33428398968
head       c6664ad4f22c25c358316aae97e7850f69279471
conclusion SUCCESS
```

All stages completed: target coverage, both fine-tunes, both W4 builds, train replay, held-out evaluation, receipt, contract assertions, and artifact upload.

### Controlled training contract

Common to both arms:

```text
cactus-needle 2.0.8
seed 0 before pinned finetune implementation
epochs 1
batch 2
lr 1e-4
LoRA rank 4 / alpha 32
6 optimizer steps / warmup 1
uniform W4 export
max_new_tokens 256
same 12 training rows
same deterministic validation row 10
```

Only intentional training difference:

```text
A cap 128  -> effective seq_len 128 -> truncated training rows [7,9]
B cap 1024 -> effective seq_len 256 -> truncated training rows []
```

### Train replay — primary measurement

```text
A 1/12 = .0833
B 1/12 = .0833
delta  = .0000
changed predictions = 0/12
```

Removing target truncation produced **no training-replay improvement and no prediction change at all** on the 12 original training queries.

### Held-out — secondary measurement

```text
A 6/24 = .250
B 6/24 = .250
delta  = .000
routing accuracy A/B = 0/18
negative-control no-call A/B = 6/6
```

One held-out prediction changed (`verify-probe-02`, expected `PROBE`): `READY -> UNKNOWN`; both predictions were wrong.

### Optimization traces

```text
                    A / 128           B / full target
step 1 loss          3.7452             3.7452
step 2 loss          3.2345             3.1537
step 3 loss          3.7264             3.7264
step 4 loss          3.4763             3.3823
step 5 loss          3.2153             3.2153
step 6 loss          3.6623             3.6623
validation loss      3.3772             3.3772
finetune wall        32.47 s            36.84 s
peak RSS             2,315,572 KiB      2,455,388 KiB
```

Loss differs only on some middle updates but converges to the same logged final/validation loss in this tiny run.

### Artifact identity and independent verification

```text
adapter A sha256 be79e53f9fb76cd355b07a98870b393f49bee072f1367890bf0745fe59e2a7f1
adapter B sha256 3cc6951e3bae7dcace0b5bec07db8dbe50406a3631ad20c7acf0f322f50a423a
W4 A sha256      2a9062cf1d2e3be3b6539e312156dea383175dd2ed3b494adc840b436534eee1
W4 B sha256      ba0e43ac574113e4b6e20affb09b56a0f7ced57d4075e97f8cfdf0c3546ce4a3
```

The downloaded artifact hashes matched `receipt.json`. Raw JSONL results were independently rescored and matched the receipt exactly.

### Scientific disposition

**FACT:** removing the two truncated training targets changed the learned adapter/artifact identity.

**FACT:** it did not improve exact train replay, held-out accuracy, routing accuracy, or negative-control behavior in this bounded one-epoch setup.

**REJECTED (bounded hypothesis):** target truncation is not supported as the dominant cause of the original microlearning failure.

**NOT CLAIMED:** truncation has no effect. Different adapter hashes and one changed held-out prediction show the training trajectory did change.

**NEXT QUESTION, NOT YET RUN:** training strength/task encoding remains the higher-value distinction. Any next experiment should vary one controlled optimization factor only after a separate specification/approval.
